include/lsst/afw/math/statistics/Framework.h

```diff
@@ -0,0 +1,615 @@
+ *   - helper (type)functions
+ *
+ * `Product`s (and `Validator`s) are composable (e.g. `Mean` depends on `Sum` and `SumWeights`).
+ * The framework collapses intermediate products together so that all products are calculated only once,
+ * in a single pass through the data, and only take up the stack space of their individual components.
+ *
+ * Each `Product` (optionally) supplies an `accumulate` (for valid values) and an `accumulateClipped` member,
+ * a `ConcreteProduct`s `accumulate` (and `accumulateClipped`) then delegates to each of its component
+ * members.
+ *
+ * All these calls are compiled away and what is left is equivalent to having only the computations required
+ * for each of the products (without dupplicates).
+ * This approach is similar, but not quite identical to expression templates.
+ */
+
+/**
+ * Set of Products
+ */
+template <typename... Ts>
+using ProductSet = std::tuple<Ts...>;
+
+/**
+ * Class to represent "no product".
+ *
+ * To be used for empty intermediary products.
+ */
+class None {};
+
+/**
+ * Type trait. If `T` is part of `Ts...`, provides the member constant value equal to `true`.
+ * Otherwise value is `false`.
+ */
+template <typename... Ts>
+struct IsIn;
+
+template <typename T, typename U>
+struct IsIn<T, U> {
+    static const bool value = std::is_same<T, U>::value;
+};
+
+template <typename T, typename U, typename... Ts>
+struct IsIn<T, U, Ts...> {
+    static const bool value = std::is_same<T, U>::value || IsIn<T, Ts...>::value;
+};
+
+/**
+ * Type trait. If `T` is part of `ProductSet<Ts...>`, provides the member constant value equal to `true`.
+ * Otherwise value is `false`.
+ */
+template <typename T, typename... Ts>
+struct IsInProductSet;
+
+template <typename T, typename... Ts>
+struct IsInProductSet<T, ProductSet<Ts...>> {
+    static const bool value = IsIn<T, Ts...>::value;
+};
+
+/**
+ * Type function to add type `T` to `ProductSet<Ts...>`.
+ *
+ * If `T` is not `None` and not already in `Ts...` then `type` is `ProductSet<T, Ts...>`,
+ * otherwise it is just `ProductSet<Ts...>`.
+ *
+ * Keeps the invariant that if `ProductSet<Ts...>` is a set of unique (and not `None`) types,
+ * so is `InsertProductSet<T, ProductSet<Ts...>>::type`.
+ */
+template <typename T, typename... Ts>
+struct ProductSetInsert;
+
+template <typename T, typename... Ts>
+struct ProductSetInsert<T, ProductSet<Ts...>> {
+    using type =
+            typename std::conditional<IsIn<T, Ts...>::value, ProductSet<Ts...>, ProductSet<T, Ts...>>::type;
+};
+
+template <typename... Ts>
+struct ProductSetInsert<None, ProductSet<Ts...>> {
+    using type = ProductSet<Ts...>;
+};
+
+/**
+ * Type function to add types `Ts...` to `ProductSet` `PS`.
+ *
+ * Keeps the invariant that if `PS` is a set of unique (and not `None`) types,
+ * so is `ProductSetInsertMultiple<PS, Ts...>::type`.
+ */
+template <typename PS, typename... Ts>
+struct ProductSetInsertMultiple;
```

Comment:
I worry that users will find it confusing to have `ProductSetInsert` work so differently than `ProductSetInsertMultiple`. `ProductSetInsert ` is templated on the contained product types, with the new type first, whereas `ProductSetInsertMultiple` is templated on the product set followed by the types to be inserted.

Could you manage with just `ProductSetInsertMultiple` and rename it to `ProductSetInsert` (or `AddProduct` or `ProductSetAdd`, as suggested above), which inserts one or more products into a `ProductSet`? If so, I think it would be less confusing.